solver/src/liquidity.rs

```diff
@@ -18,6 +20,7 @@ pub mod uniswap;
 pub enum Liquidity {
     Limit(LimitOrder),
     ConstantProduct(ConstantProductOrder),
+    WeightedProduct(WeightedProductOrder),
```

Comment:
Would a 50:50 balancer pool still be modelled as a constant product order? In this case maybe we don't need the WeightedPool type until the solver can actually handle e.g. 80:20 pools?

Reply:
At this moment it was the plan to be putting all balancer pools into the weighted pool section since their settlement encoding is different.

Reply:
AFAIU the current plan is to pass weighted pools to the solver as is, and internally they will treat it like a Uniswap pool if the weights are the same.

Reply:
Perhaps to indicate this better we should actually use the term balancer and uniswap pools on the inner types.

Reply:
Also, this is the abstract solver liquidity type that goes to all solvers, including the baseline solver which should be able to handle 80:20 pools (since it works by computing output amounts, for which we already have the math).

> since their settlement encoding is different

This should be handled by the `SettlementHandling` implementation, which in theory could also be made to work with Balancer pools.

That being said, I do believe it is correct here to distinguish between weighted and constant product pools.

Reply:
> Perhaps to indicate this better we should actually use the term balancer and uniswap pools on the inner types

I think the names are correct. The interaction encoding is done by the `SettlementHandling` implementation which already has [names based on the protocols they are encoding for](https://github.com/gnosis/gp-v2-services/blob/main/solver/src/liquidity/uniswap.rs).

Reply:
That's all fine and dandy, but I am no longer sure if you are in favour of Felix's suggestion to assign 50:50 balancer pools to constant product orders. It is my impression that this was not the plan.

Reply:
I am for leaving it as `WeightedPool`.

Reply:
Just want to highlight that the fact that their settlement logic is different, doesn't mean they need to be separate liquidity sub-types. Each liquidity type has a dynamic "settlement_handling" which encodes how this specific instance gets translated into interactions. Two different `ConstantProductOrder` orders (e.g Uniswap, Sushiswap) can have different settlement_handling (one settles on Uni the other one on Sushi). Similarly, both EOA as well as 0x orders can be modelled with the LimitOrder subtype (despite very different settlement logic).

The concern about rounding makes sense. I think midterm it would be nice if this enum is not overlapping and since a `WeightedProductOrder` includes `ConstantProductOrder` they could be modelled as one subtype.

Reply:
Maybe @fedgiac knows if the Uniswap v2 math is equivalent to the Balancer v2 math (for 50:50 pools) or if there can be rounding discrepancies

Reply:
Given that balancer uses their own floating point implementation on the contract level, I think we can be quite confident that, although the formulas may boil down to the same thing in overlapping circumstances, the arithmetic is bound to differ by even a marginal little bit on some corner case. Would also be nice to hear from Federico though!

Reply:
One place to look for a difference is a uniswap pool with a token having more than 18 decimals. Balancer does not support tokens with this property but this would be a place to look for a clear difference in valuations. Another place that might be more relevant is just a two token pool where one token has 18 decimals and the other has 0 which would be possible in both types of pools. We could set the pools to have the exact same reserves and compute the in given out or vice versa for both. See what comes out.

Reply:
If this is the case, then I'd question if the separation into ConstantProduct and WeightedProduct order makes sense. If we really need a different liquidity type per protocol (because of rounding details in the concrete implementations) then we should probably call the Liquidity types explicitly after their protocol. If however the math can be approximated by one another then the abstraction into types of functions (Limit, WeightedPool, ArithmeticMeanPool) makes sesne.

Reply:
I am currently experimenting with an `AmmOrder<T>` with a Liquidity enum that looks like this:

```rust
pub struct AmmOrder<T> {
    pub reserves: HashMap<H160, T>,
    pub fee: Ratio<u32>,
    pub settlement_handling: Arc<dyn SettlementHandling<Self>>,
}
```

```rust
pub enum Liquidity {
    Limit(LimitOrder),
    Uniswap(AmmOrder<u128>),
    Balancer(AmmOrder<PoolTokenState>),
}
```

Is this something that looks a bit like what you are thinking.

I actually think this might work out very well, but does require some changes.

Reply:
Unfortunately this is goingo to be quite an invasive change of structure given how much we rely on the `ConstantProductOrder` struct as it currently is. I don't belive it is worth the abstraction.